### PR TITLE
Reuse an event start and BytesStart when writing the svg

### PR DESCRIPTION
**Pull request by meven**
cc https://github.com/jonhoo/inferno/issues/41

@JordiChauzi you can ignore the PR, it is meant to get merged after your `full_color_support` branch gets merged.

cc @jonhoo read ahead review

### CODE REVIEW

**Comment by jonhoo** [comment]
@meven this is a good start, but note that `filled_rectangle` *also* creates an element!

**Comment by meven** [comment]
> @meven this is a good start, but note that `filled_rectangle` _also_ creates an element!

Thanks

File: src/flamegraph/mod.rs [hunk 1]
```diff
@@ -205,7 +205,8 @@ where
 
     // structs to reuse accross loops to avoid allocations
     let mut event_start = Event::Start({ BytesStart::owned_name("g") });
-    let mut event_start_a = Event::Start({BytesStart::owned_name("a")});
+    let mut event_start_a = Event::Start({ BytesStart::owned_name("a") });
+    let mut event_empty_filled_rect = Event::Empty(BytesStart::owned_name("rect"));
```
Comment (jonhoo) [inline src/flamegraph/mod.rs]: I think I would just name these `cache_X`:
```rust
let mut cache_g = Event::Start({ BytesStart::owned_name("g") });
let mut cache_a = Event::Start({ BytesStart::owned_name("a") });
let mut cache_rect = Event::Empty(BytesStart::owned_name("rect"));
```

File: src/flamegraph/mod.rs [hunk 2]
```diff
@@ -424,22 +460,26 @@ fn filled_rectangle<W: Write>(
...
+        bytes_start.extend_attributes(args!(
+            "x" => &buffer[x],
+            "y" => &buffer[y],
+            "width" => &buffer[width],
+            "height" => &buffer[height],
+            "fill" => &buffer[color]
+        ));
+    }
+    svg.write_event(&svg_event_empty)
 }
```
Comment (jonhoo) [inline src/flamegraph/mod.rs:464]: We may want
```rust
} else {
    unreachable!("cache wrapper was of wrong type: {:?}", svg_event_empty);
}
```
And maybe also in the other places where we use this pattern?

File: src/flamegraph/mod.rs [hunk 3]
```diff
@@ -424,22 +460,26 @@ fn filled_rectangle<W: Write>(
     y1: usize,
     y2: usize,
     color: (u8, u8, u8),
+    svg_event_empty: &mut Event,
```
Comment (jonhoo) [inline src/flamegraph/mod.rs]: There are so many arguments here now. Maybe make them a struct?

Comment (meven) [reply src/flamegraph/mod.rs]: I picked up the idea, the code looks better indeed.

File: src/flamegraph/mod.rs (hunk 2 above)
Comment (meven) [reply src/flamegraph/mod.rs:464]: I thought about

```rust
        if href_is_some {
          svg.write_event(Event::End(BytesEnd::borrowed(b"a")))?;
        }
        svg.write_event(Event::End(BytesEnd::borrowed(b"g")))?;
```
But I am guessing the compiler is good enough to "constantify" those.

**Pull request closed without merging**